modules/core/src/main/scala/parser.scala

```diff
@@ -312,71 +316,145 @@ object GraphQLParser {
     } yield Ast.VariableDefinition(v.name, tpe, dv)
 
   lazy val Variable: Parser[Ast.Value.Variable] =
-    keyword("$") ~> Name.map(Ast.Value.Variable.apply)
+    keyword("$") *> Name.map(Ast.Value.Variable.apply)
 
   lazy val DefaultValue: Parser[Ast.Value] =
-    keyword("=") ~> Value
+    keyword("=") *> Value
 
-  lazy val Type: Parser[Ast.Type] =
-    NonNullType.widen[Ast.Type] | ListType.widen | NamedType.widen
+  lazy val Type: Parser[Ast.Type] = recursive[Ast.Type] { rec =>
 
-  lazy val NamedType: Parser[Ast.Type.Named] =
-    Name.map(Ast.Type.Named.apply)
+    lazy val ListType: Parser[Ast.Type.List] =
+      squareBrackets(rec).map(Ast.Type.List.apply)
 
-  lazy val ListType: Parser[Ast.Type.List] =
-    squareBrackets(Type).map(Ast.Type.List.apply)
+    val namedMaybeNull: Parser[Ast.Type] = (NamedType ~ keyword("!").?).map {
+      case (t, None) => t
+      case (t, _) => Ast.Type.NonNull(Left(t))
+    }
 
-  lazy val NonNullType: Parser[Ast.Type.NonNull] =
-    (NamedType <~ keyword("!")).map(a => Ast.Type.NonNull(Left(a))).widen |
-    (ListType  <~ keyword("!")).map(a => Ast.Type.NonNull(Right(a))).widen
+    val listMaybeNull: Parser[Ast.Type] = (ListType ~ keyword("!").?).map {
+      case (t, None) => t
+      case (t, _) => Ast.Type.NonNull(Right(t))
+    }
 
-  lazy val Directives =
+    namedMaybeNull | listMaybeNull
+  }
+
+  lazy val NamedType: Parser[Ast.Type.Named] =
+    Name.map(Ast.Type.Named.apply)
+
+  lazy val Directives: Parser0[List[Ast.Directive]] =
     many(Directive)
 
   lazy val Directive: Parser[Ast.Directive] =
-    keyword("@") ~> (Name, opt(Arguments)).mapN((n, ods) => Ast.Directive(n, ods.orEmpty))
+    keyword("@") *> (Name ~ opt(Arguments)).map { case (n, ods) => Ast.Directive(n, ods.orEmpty)}
 
   lazy val Name: Parser[Ast.Name] = {
     val initial    = ('A' to 'Z').toSet ++ ('a' to 'z').toSet + '_'
```

Comment:
nit: you don't need the `.toSet` on the `('a' to 'z')` you just need an iterable